### PR TITLE
🔒 M-03 - Add Authorization Control to Fallback Function

### PR DESCRIPTION
**Pull request by Aboudjem**
#### M-03. Anyone can call the fallbackFunction because of missing authorization control

- **Issue**: Lack of authorization control in the fallback function.
- **Affected Functions**: `fallback` function in `ModuleManager`.
- **Fix**: Implement proper authorization control to ensure only authorized entities can invoke it.

### CODE REVIEW

File: contracts/Nexus.sol [hunk 1]
```diff
@@ -137,15 +138,22 @@ contract Nexus is INexus, BaseAccount, ExecutionHelper, ModuleManager, UUPSUpgra
     function executeFromExecutor(
         ExecutionMode mode,
         bytes calldata executionCalldata
-    ) external payable onlyExecutorModule withHook withRegistry(msg.sender, MODULE_TYPE_EXECUTOR) returns (bytes[] memory returnData) {
+    )
```
Comment (livingrockrises) [inline contracts/Nexus.sol:141]: can we not touch linting as far as this PR goes?

File: contracts/base/BaseAccount.sol [hunk 2]
```diff
@@ -23,14 +25,10 @@ import { IBaseAccount } from "../interfaces/base/IBaseAccount.sol";
 /// @author @filmakarov | Biconomy | filipp.makarov@biconomy.io
 /// @author @zeroknots | Rhinestone.wtf | zeroknots.eth
 /// Special thanks to the Solady team for foundational contributions: https://github.com/Vectorized/solady
-contract BaseAccount is IBaseAccount {
+contract BaseAccount is Storage, IBaseAccount {
```
Comment (livingrockrises) [inline contracts/base/BaseAccount.sol:28]: ??

**Comment by livingrockrises** [comment]
@Aboudjem 

what is done specific to this?
> Fix: Implement proper authorization control to ensure only authorized entities can invoke it

I don't see changes related to it. we also discussed fallback handler should implement it's own auth control and we are gonna make changes on the EIP PR. cc @filmakarov 

can you close this PR and redo it if only there are additional test cases. bunch of linting makes it very hard to review..

**Comment by VGabriel45** [comment]
Should this be a remediation for this https://codehawks.cyfrin.io/c/2024-07-biconomy/s/179 ?

File: contracts/base/BaseAccount.sol (hunk 2 above)
Comment (Aboudjem) [reply contracts/base/BaseAccount.sol:28]: I moved the _ENTRYPOINT to storage, as it needs to be used by both BaseAccount and ModuleManager for the new `onlyAuthorized`

**Comment by livingrockrises** [comment]
> Should this be a remediation for this https://codehawks.cyfrin.io/c/2024-07-biconomy/s/179 ?

No, anyone should be able to invoke fallback handlers.
the actual access control (if required) should be implemented in the handlers themselves..

File: contracts/base/BaseAccount.sol (hunk 2 above)
Comment (livingrockrises) [reply contracts/base/BaseAccount.sol:28]: why would you do that?
it's not a storage
it's immutable : address internal immutable _ENTRYPOINT;

what's new onlyAuthorized for?

**Comment by Aboudjem** [comment]
> @Aboudjem
> 
> what is done specific to this?
> 
> > Fix: Implement proper authorization control to ensure only authorized entities can invoke it
> 
> I don't see changes related to it. we also discussed fallback handler should implement it's own auth control and we are gonna make changes on the EIP PR. cc @filmakarov
> 
> can you close this PR and redo it if only there are additional test cases. bunch of linting makes it very hard to review..

Implement proper authorization control in the fallback function to ensure that only authorized entities can invoke it. This can be achieved by adding a modifier to check the sender's authorization before routing the call to the fallback handler. The existing onlyEntryPointOrSelf modifier could be used or an new modifier also including executorModuls might be appropriate.

File: contracts/base/BaseAccount.sol (hunk 2 above)
Comment (Aboudjem) [reply contracts/base/BaseAccount.sol:28]: onlyAuthorized is checking msg.sender is EP, self or Executor while other modifier are exclusively checking EP EP+self or Executor only

if the naming is a problem we can change it

**Comment by livingrockrises** [comment]
> > @Aboudjem
> > what is done specific to this?
> > > Fix: Implement proper authorization control to ensure only authorized entities can invoke it
> > 
> > 
> > I don't see changes related to it. we also discussed fallback handler should implement it's own auth control and we are gonna make changes on the EIP PR. cc @filmakarov
> > can you close this PR and redo it if only there are additional test cases. bunch of linting makes it very hard to review..
> 
> `Implement proper authorization control in the fallback function to ensure that only authorized entities can invoke it. This can be achieved by adding a modifier to check the sender's authorization before routing the call to the fallback handler. The existing onlyEntryPointOrSelf modifier could be used or an new modifier also including executorModuls might be appropriate.`
> 
> At this point, the entrypoint and the onlyEntrypoint or onlySelfOrEntrypoint are not visible, any suggestion?
> 
> Yes my husky script messed a bit with the linter

dont think any of this change makes sense.

**Comment by Aboudjem** [comment]
> > > @Aboudjem
> > > what is done specific to this?
> > > > Fix: Implement proper authorization control to ensure only authorized entities can invoke it
> > > 
> > > 
> > > I don't see changes related to it. we also discussed fallback handler should implement it's own auth control and we are gonna make changes on the EIP PR. cc @filmakarov
> > > can you close this PR and redo it if only there are additional test cases. bunch of linting makes it very hard to review..
> > 
> > 
> > `Implement proper authorization control in the fallback function to ensure that only authorized entities can invoke it. This can be achieved by adding a modifier to check the sender's authorization before routing the call to the fallback handler. The existing onlyEntryPointOrSelf modifier could be used or an new modifier also including executorModuls might be appropriate.`
> > At this point, the entrypoint and the onlyEntrypoint or onlySelfOrEntrypoint are not visible, any suggestion?
> > Yes my husky script messed a bit with the linter
> 
> dont think any of this change makes sense.

then close the PR

**Pull request closed without merging**